graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/rest/PipelineResource.java

```diff
@@ -250,6 +272,111 @@ public PipelineSource update(@ApiParam(name = "id") @PathParam("id") String id,
+    @Path("/routing")
+    @PUT
+    @AuditEvent(type = PipelineProcessorAuditEventTypes.PIPELINE_UPDATE)
+    public PipelineSource routing(@ApiParam(name = "body", required = true) @NotNull RoutingRequest request) throws NotFoundException {
+        checkPermission(RestPermissions.STREAMS_EDIT, request.streamId());
+        checkPermission(PipelineRestPermissions.PIPELINE_RULE_CREATE);
+
+        String streamName;
+        try {
+            final Stream stream = streamService.load(request.streamId());
+            streamName = stream.getTitle();
+        } catch (NotFoundException e) {
+            throw new InternalServerErrorException(f("Unable to load stream %s", request.streamId()), e);
+        }
+
+        RuleDao ruleDao = createRoutingRule(request, streamName);
+        PipelineDao pipelineDao;
+        try {
+            pipelineDao = pipelineService.loadByName(GL_INPUT_ROUTING_PIPELINE);
+        } catch (NotFoundException e) {
+            // Create pipeline with first rule
+            return createRoutingPipeline(ruleDao);
+        }
+
+        // Add rule to existing pipeline
+        PipelineSource pipelineSource = PipelineSource.fromDao(pipelineRuleParser, pipelineDao);
+        final List<String> rules0 = pipelineSource.stages().get(0).rules();
+        if (rules0.stream().filter(ruleRef -> ruleRef.equals(ruleDao.title())).findFirst().isEmpty()) {
+            rules0.add(ruleDao.title());
+            PipelineSource updated = pipelineSource.toBuilder()
+                    .source(createPipelineString(pipelineSource))
+                    .build();
+            update(pipelineDao.id(), updated);
+        } else {
+            log.info(f("Routing for input <%s> already exists - skipping", request.inputId()));
+        }
+
+        return pipelineSource;
+    }
+
+    private PipelineSource createRoutingPipeline(RuleDao ruleDao) {
+        List<StageSource> stages = java.util.List.of(StageSource.create(
+                0, Stage.Match.EITHER, java.util.List.of(ruleDao.title())));
+        final PipelineSource pipelineSource = PipelineSource.builder()
+                .title(GL_INPUT_ROUTING_PIPELINE)
+                .description("GL generated pipeline")
+                .source("pipeline \"" + GL_INPUT_ROUTING_PIPELINE + "\"\nstage 0 match either\nrule \"" + ruleDao.title() + "\"\nend")
+                .stages(stages)
+                .build();
+        final PipelineSource parsedSource = createFromParser(pipelineSource);
+
+        Set<String> pipelineIds = new HashSet<>();
+        pipelineIds.add(parsedSource.id());
+        final PipelineConnections pipelineConnections = PipelineConnections.create(null, DEFAULT_STREAM_ID, pipelineIds);
+        connectionsService.save(pipelineConnections);
+
+        return parsedSource;
+    }
+
+    private RuleDao createRoutingRule(RoutingRequest request, String streamName) {
+        String ruleName = "route_" + request.inputId() + "_to_" + streamName;
```

Comment:
According to this text:
> When the user chooses a stream, at this stage we automatically create a rule in Stage 0 of this pipeline called “Route to [Stream name]”. 

we should call the rule “Route to [Stream name]”. But I like the idea of including an identifier from the input. I would prefer the name of the input, as otherwise it is too cryptic in my opinion. @tellistone what is your opinion on this.

Reply:
We need the input ID to disambiguate, in case the user decides to route multiple inputs to the same stream.